spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/BaseConfigServicePropertySourceLocator.java

```diff
@@ -0,0 +1,140 @@
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.HttpServerErrorException;
+
+/**
+ * 
+ * @author Jeffrey Nelson
+ *
+ * @param <C> type of Context class used
+ */
+public abstract class BaseConfigServicePropertySourceLocator<C extends BaseConfigServicePropertySourceLocator.Context>
+		implements PropertySourceLocator {
+
+	private static Log logger = LogFactory.getLog(BaseConfigServicePropertySourceLocator.class);
+
+	private final ConfigClientProperties defaults;
+	private final Class<C> contextClass;
+
+	public BaseConfigServicePropertySourceLocator(ConfigClientProperties defaults, Class<C> contextClass) {
+		this.defaults = defaults;
+		this.contextClass = contextClass;
+	}
+
+	@Override
+	@Retryable(interceptor = "configServerRetryInterceptor")
+	public org.springframework.core.env.PropertySource<?> locate(org.springframework.core.env.Environment environment) {
+		C context = newContext(environment);
+		fillContext(context, environment);
+
+		CompositePropertySource composite = new CompositePropertySource("configService");
+		Exception error = null;
+		String errorBody = null;
+		logger.info("Fetching config from server");
+		try {
+			String[] labels = new String[] { "" };
+			if (StringUtils.hasText(context.getConfigClientProperties().getLabel())) {
+				labels = StringUtils.commaDelimitedListToStringArray(context.getConfigClientProperties().getLabel());
+			}
+			// Try all the labels until one works
+			for (String label : labels) {
+				context.setLabel(label.trim());
+				Environment result = getRemoteEnvironment(context);
+				if (result != null) {
+					logger.info(String.format("Located environment: name=%s, profiles=%s, label=%s, version=%s",
+							result.getName(), result.getProfiles() == null ? "" : Arrays.asList(result.getProfiles()),
+							result.getLabel(), result.getVersion()));
+
+					for (PropertySource source : result.getPropertySources()) {
+						@SuppressWarnings("unchecked")
+						Map<String, Object> map = (Map<String, Object>) source.getSource();
+						composite.addPropertySource(new MapPropertySource(source.getName(), map));
+					}
+					return composite;
+				}
+			}
+		} catch (HttpServerErrorException e) {
+			error = e;
+			if (MediaType.APPLICATION_JSON.includes(e.getResponseHeaders().getContentType())) {
+				errorBody = e.getResponseBodyAsString();
+			}
+		} catch (Exception e) {
+			error = e;
+		}
+		if (context.getConfigClientProperties() != null && context.getConfigClientProperties().isFailFast()) {
+			throw new IllegalStateException(
+					"Could not locate PropertySource and the fail fast property is set, failing", error);
+		}
+		logger.warn("Could not locate PropertySource: "
+				+ (errorBody == null ? error == null ? "label not found" : error.getMessage() : errorBody));
+		return null;
+
+	}
+	
+	protected C newContext(org.springframework.core.env.Environment environment) {
+		try {
+			C context = contextClass.newInstance();
```

Comment:
OK, another option is to make newContext abstract and require child classes to implement. 

I did it this way because I see the Context class as a simple POJO that should always have default ctor. And in some case where that won't work, the method is already protected and could be overridden.